feat: new pg-pg migration doc for CH cloud (#6113)
ClickHouse/clickhouse-docs: docs/cloud/managed-postgres/migrations/clickhouse-cloud.md

import serviceCard from '@site/static/images/managed-postgres/pgpg/servicecard.png';
import startImport from '@site/static/images/managed-postgres/pgpg/startimport.png';
import tablePicker from '@site/static/images/managed-postgres/pgpg/tablepicker.png';

# Migrate to Managed Postgres using ClickHouse Cloud {#migrate-managed-postgres}

<PrivatePreviewBadge link="https://clickhouse.com/cloud/postgres" galaxyTrack={true} slug="migration-guide-clickhouse-cloud" />

ClickHouse Cloud includes a built-in import wizard that migrates your external PostgreSQL database into a Managed Postgres service. The wizard handles the source connection, schema export and import, replication settings, and table selection in five guided steps.

## Prerequisites {#prerequisites}

- Access to your source PostgreSQL database with a user that has replication privileges.
- A ClickHouse Managed Postgres service as the migration target. If you don't have one yet, see the [quickstart](../quickstart).
- `pg_dump` and `psql` installed on your local machine. Both ship with the standard PostgreSQL client tools.

## Considerations before migrating {#considerations}

> - **DDL propagation**: continuous replication (CDC) captures DML operations and `ADD COLUMN`. Other DDL changes such as `DROP COLUMN` and `ALTER COLUMN` are not propagated and must be applied manually on the target.

Reviewer: ```suggestion
- **DDL propagation**: continuous replication (CDC) captures DML operations and `ADD COLUMN`. Other DDL changes such as `DROP COLUMN` and `ALTER COLUMN` aren't propagated and must be applied manually on the target.
```

- **Foreign key constraints**: to prevent ingestion from being blocked by foreign key checks, you'll temporarily set `session_replication_role = replica` on the target role. This is covered in step 3 below.

## Step 1: Connect to your source database {#step-1-connect}

Open the [ClickHouse Cloud console](https://clickhouse.cloud) and select your Managed Postgres service.

<Image img={serviceCard} alt="Managed Postgres service card in the ClickHouse Cloud services list" size="lg" border />

In the left sidebar, click **Data sources**.

<Image img={overview} alt="Data sources entry in the Managed Postgres service sidebar" size="lg" border />

Click **Start import**.

<Image img={startImport} alt="Data sources page with Start import button" size="lg" border />

Fill in the connection details for your source PostgreSQL database: host, port, username, password, and database name. Enable **TLS** if your source requires it.

> If you require a private connection to your source database, you can opt for **SSH tunneling** and provide the necessary SSH details. This allows the migration to securely connect to databases that are not publicly accessible.

Reviewer: ```suggestion
If you require a private connection to your source database, you can opt for **SSH tunneling** and provide the necessary SSH details. This allows the migration to securely connect to databases that aren't publicly accessible.
```

<Image img={nextImport} alt="Step 3: psql command for schema import" size="lg" border />

```shell
psql \
  -h <target_host> \
  -p 5432 \
  -U <target_user> \
  -d <target_database> \
  -f pg.sql
```

<Image img={psqlImport} alt="Terminal output after running psql schema import" size="lg" border />

After the schema is applied, set `session_replication_role` to `replica` on the target role so that foreign key constraints don't block ingestion:

```sql
ALTER ROLE <target_role> SET session_replication_role TO 'replica';
```

<Image img={alterRole} alt="ALTER ROLE command setting session_replication_role to replica" size="lg" border />

Click **Next**.

## Step 4: Configure ingestion settings {#step-4-ingestion-settings}

Specify the publication to use for logical replication. If you leave this blank, a publication is created automatically.

Expand **Advanced replication settings** to tune throughput:

| Setting | Default | Description |
|---|---|---|
| Sync interval (seconds) | 10 | How frequently the replication slot is polled |
| Parallel threads for initial load | 4 | Number of threads for the bulk copy phase |

> | Pull batch size | 100000 | Rows fetched per replication batch |

Reviewer: ```suggestion
| Pull batch size | 100,000 | Rows fetched per replication batch |
```

```sql
ALTER ROLE <target_role> SET session_replication_role TO 'origin';
```

**Reset sequences.** Align sequences with the current maximum values in each table:

```sql
DO $$
DECLARE r RECORD;
BEGIN
    FOR r IN
        SELECT
            n.nspname AS schema_name,
            c.relname AS table_name,
            a.attname AS column_name,
            pg_get_serial_sequence(format('%I.%I', n.nspname, c.relname), a.attname) AS seq_name
        FROM pg_class c
        JOIN pg_namespace n ON n.oid = c.relnamespace
        JOIN pg_attribute a ON a.attrelid = c.oid
        WHERE c.relkind = 'r'
            AND a.attnum > 0
            AND NOT a.attisdropped
            AND n.nspname NOT IN ('pg_catalog', 'information_schema')
    LOOP
        IF r.seq_name IS NOT NULL THEN
            EXECUTE format(
                'SELECT setval(%L, COALESCE((SELECT MAX(%I) FROM %I.%I), 0) + 1, false)',
                r.seq_name, r.column_name, r.schema_name, r.table_name
            );
        END IF;
    END LOOP;
END $$;
```

**Cut over application traffic.** Point reads and writes to your Managed Postgres service and monitor for errors, constraint violations, and replication health.

> **Clean up.** Once you're satisfied, delete the migration from **Data sources**. If you used CDC, drop the replication slot from the source to free resources:

Reviewer: ```suggestion
**Clean up.**  Once you've cut over and confirmed the new service is healthy, delete the migration from **Data sources**. If you used CDC, drop the replication slot from the source to free resources:
```
